### PR TITLE
Fix indexScan rule when compare INT with Double

### PR DESCRIPTION
**Pull request by Aiee**
1. Fix crash caused by comparing different numerical types in the index scan. Issue: fixes https://github.com/vesoft-inc/nebula-graph/issues/721

TODO:
Fix edges cases failure due to the float precision.

### CODE REVIEW

File: src/optimizer/OptimizerUtils.cpp [hunk 1]
```diff
@@ -32,26 +32,44 @@ Value OptimizerUtils::boundValue(const meta::cpp2::ColumnDef& col,
...
+                if (!const_cast<Value&>(v).toInt().second) {
+                    return Value::kNullBadType;
+                }
+                boundVar = const_cast<Value&>(v).toInt().first;
+            } else {
+                boundVar = v.getInt();
+            }
+
+            if (boundVar == std::numeric_limits<int64_t>::max()) {
+                return Value(boundVar);
```
Comment (bright-starry-sky) [inline src/optimizer/OptimizerUtils.cpp:48]: I don't think it's a good idea. this leads to the logic more complex in future.
If you want to support value comparison between different type, you’d better cast the data type in validator .

File: src/optimizer/OptimizerUtils.cpp [hunk 2]
```diff
@@ -32,26 +32,44 @@ Value OptimizerUtils::boundValue(const meta::cpp2::ColumnDef& col,
...
+        case Value::Type::FLOAT: {
+            double_t boundVar;
+            if (v.isInt()) {
+                boundVar = const_cast<Value&>(v).toFloat().first;
+            } else {
+                boundVar = v.getFloat();
+            }
+
+            if (boundVar > 0.0) {
+                if (boundVar == std::numeric_limits<double_t>::max()) {
```
Comment (bright-starry-sky) [inline src/optimizer/OptimizerUtils.cpp:62]: ditto

File: src/optimizer/OptimizerUtils.cpp [hunk 3]
```diff
@@ -181,37 +199,46 @@ Value OptimizerUtils::boundValueWithLT(const meta::cpp2::ColumnDef& col, const V
...
             }
             std::vector<unsigned char> bytes(v.getStr().begin(), v.getStr().end());
             bytes.resize(*col.get_type().get_type_length());
-            for (size_t i = bytes.size();; i--) {
-                if (i > 0) {
-                    if (bytes[i-1]-- != 0) break;
-                } else {
-                    return Value(std::string(*col.get_type().get_type_length(), '\0'));
-                }
-            }
```
Comment (bright-starry-sky) [inline src/optimizer/OptimizerUtils.cpp:214]: why ?

File: src/optimizer/rule/IndexScanRule.cpp [hunk 4]
```diff
@@ -246,26 +251,30 @@ Status IndexScanRule::boundValue(const FilterItem& item,
...
             // where c >= 1 and c >= 2 , 2 should be valid.
             if (begin == Value()) {
-                begin = val;
+                begin = v;
             } else {
-                begin = val < begin ? begin : val;
+                begin = v < begin ? begin : v;
             }
             break;
         }
```
Comment (bright-starry-sky) [inline src/optimizer/rule/IndexScanRule.cpp:263]: What is the purpose?

File: src/optimizer/test/IndexBoundValueTest.cpp [hunk 5]
```diff
@@ -36,7 +36,7 @@ TEST(IndexBoundValueTest, StringTest) {
     EXPECT_EQ(maxStr, OptimizerUtils::boundValue(col, OP::GREATER_THAN, Value(maxStr)).getStr());
 
     retVal = OptimizerUtils::boundValue(col, OP::LESS_THAN, Value("aa")).getStr();
-    expected = {'a', '`', 255, 255, 255, 255, 255, 255};
+    expected = {'a', 'a', '\0', '\0', '\0', '\0', '\0', '\0'};
```
Comment (bright-starry-sky) [inline src/optimizer/test/IndexBoundValueTest.cpp:39]: why?

File: src/optimizer/test/IndexBoundValueTest.cpp [hunk 6]
```diff
@@ -48,7 +48,7 @@ TEST(IndexBoundValueTest, StringTest) {
     {
         auto actual = "ABCDEFGHIJKLMN";
         auto expectGT = "ABCDEFGI";
-        auto expectLT = "ABCDEFGG";
+        auto expectLT = "ABCDEFGH";
```
Comment (bright-starry-sky) [inline src/optimizer/test/IndexBoundValueTest.cpp:51]: why?

File: src/optimizer/rule/IndexScanRule.cpp [hunk 7]
```diff
@@ -228,13 +228,18 @@ Status IndexScanRule::boundValue(const FilterItem& item,
                                  const meta::cpp2::ColumnDef& col,
                                  Value& begin, Value& end) const {
     auto val = item.value_;
-    if (val.type() != graph::SchemaUtil::propTypeToValueType(col.type.type)) {
+    // Allow different numeric types to compare
+    if (val.isNumeric() &&
+        (graph::SchemaUtil::propTypeToValueType(col.type.type) == Value::Type::INT ||
+         graph::SchemaUtil::propTypeToValueType(col.type.type) == Value::Type::FLOAT)) {
+        // skip the check
+    } else if (val.type() != graph::SchemaUtil::propTypeToValueType(col.type.type)) {
```
Comment (bright-starry-sky) [inline src/optimizer/rule/IndexScanRule.cpp:236]: Restore it, or DCHECK

**Review by bright-starry-sky — COMMENTED** [review]
I've only checked part of it, and seems we need to discuss the design

**Comment by Aiee** [comment]
After offline discussion with @bright-starry-sky, it seems there are other approaches that solve the same problem in a less complicated way.

A refactored PR has been submitted: https://github.com/vesoft-inc/nebula-graph/pull/784

**Pull request closed without merging**